### ISSUE TITLE
Starting a new connection that resumes from a snapshotted message ID

### ISSUE THREAD

**Issue by KrzysFR** [SignalR/SignalR#482]
TL/DR:

Being able to get a snapshot of the latest message Id in the message bus, pass it the the JavaScript code, and instruct the $.connection to resume from this id when connecting (like it would do on a reconnect), and not from the  current ID at the time of /signalr/negotiate or /signalr/connect.

I think I need to explain the problem I am trying to solve, in order to better understand what I mean.
### Scenario:

I have a situation where I need to synchronize a view of a collection of items being modified on a server by background workers (add, update, deleted), to a View in the user's web browser.  I am using SignalR to broadcast "onChange" events from the collection to the clients, and a client library like KnockoutJS to update the view in the DOM of the page.

There is a problem during the page load, because I would like to be able to create a snapshot of the collection inside the MVC controller, pass it to the View, and have it embed the snapshot as a JSON blob inside the page sent to the browser. This way the page can load instantly, without having to do any AJAX request to fetch the initial data (request that would happen only after every script/css is loaded).

When doing that, there is a  gap between the snapshot and when SignalR has negotiated the connection, where I will loose "onChange" events. This means that the view in the browser would not be accurate. With a busy page (lots of scripts / css to load) and lousy internet connection, the gap can grow up to 500ms or even 1 sec.

Also note that even when doing a post-load ajax request to get the data, there could be a gap if the ajax request completes BEFORE the SignalR connection is established. The only "safe" way would be to wait for the $.connection.hub.start() to finish before fetching the data via AJAX.
### Possible solution:

All the onChange events sent after the snapshot of the collection, and before the SignalR connection is established, will stay in the Message Bus for at least 30 sec (if i'm not mistaken, at least with the InProcessMessageBus).

If the MVC controller action could get a snapshot of the latest message ID, and pass it to the view (using the ViewBag for ex.), and if we could pass this ID to the $.connection object, maybe it could negotiate/connect, but also tell the SignalR server to "resume" from this ID. This way, the Javascript code would receive all pending events since the snapshot.

This would be almost like a Reconnect, where the client would catch up to all msg sent while he was reconnecting.

Also, this would give us a 30 second window for the negotiate/connect to happen.
### Sample Code:

In the controller somewhere:

```
   lock(SomeGlobalLock)
   {
       var snapshot = GetGlobalSnapshot();
       ViewData["SignalR_Id"] = SignalRContext.GetSnapshottedMessageId();
    }
```

and in the JavaScript:

```
   $.connection.resumeFromMessageId = '@ViewData["SignalR_Id"]';
   $.connection.hub.start();
```
### Possible Problems:

With web farms and load balancing, the MVC controller could execute on one server, but the SignalR connection could happen on another server. So the snapshotted message ID would need to be "global" (probably already handled with scale out). Maybe the code could create an empty "marker" message just in order to get a valid ID ?

Since obtaining the ID could take some time with ScaleOut (going out to a Redis or SQL server to generate the id), the method would need to be async (and return a Task<string> ?).

There would need to be a way to handle the case where the client wants to resume from a stale message id (the servers have rebooted since then), or if the id is too old (already been cleaned up).
